### PR TITLE
Triangulation

### PR DESCRIPTION
**Pull request by robustov**
Added triangulation and normal evaluation methods. Added some unit tests in test_core.cpp

## Summary by Sourcery

Add mesh face triangulation and automatic vertex normal computation based on triangulated geometry.

New Features:
- Triangulate polygonal mesh faces into internal triangle representations while validating vertex indices.
- Automatically compute and normalize per-vertex normals when no normals are provided.

Enhancements:
- Expose triangulated face count via Mesh::face_count and store triangulated faces separately from original faces.

Tests:
- Add unit tests covering mesh triangulation for various polygon sizes and edge cases.
- Add unit tests validating normal preservation, computed normals, invalid index handling, and normalization behavior.
- Update IO tests to compare expected face counts against the number of triangulated faces derived from original faces.

### CODE REVIEW

File: src/core/Mesh.cpp [hunk 1]
```diff
@@ -1,25 +1,79 @@
...
+                    triangulated_faces.push_back({{face[0].vi, face[i].vi, face[i + 1].vi}});
+                }
+            }
+        }
+    }
+
     void Mesh::load_texture(const std::string& filename) {
-        std::cout << "Loading " << filename << '\n';
         texture_filename = filename;
-        throw std::runtime_error("Not implemented yet!");
```
Comment (yel-gar) [inline src/core/Mesh.cpp]: Просьба пока это оставить как есть

File: src/core/Mesh.cpp [hunk 2]
```diff
@@ -1,25 +1,79 @@
...
+            if (face.size() < 3) {
+                continue;
+            }
+
+            if (face.size() == 3) {
+                triangulated_faces.push_back({{face[0].vi, face[1].vi, face[2].vi}});
+            } else if (face.size() > 3) {
+                for (size_t i = 1; i < face.size() - 1; ++i) {
+                    triangulated_faces.push_back({{face[0].vi, face[i].vi, face[i + 1].vi}});
+                }
```
Comment (yel-gar) [inline src/core/Mesh.cpp]: Чисто по стилю я бы сделал без доп ветвления, но не принципиально

```suggestion
            if (face.size() == 3) {
                triangulated_faces.push_back({{face[0].vi, face[1].vi, face[2].vi}});
                return;
            }

            for (size_t i = 1; i < face.size() - 1; ++i) {
                triangulated_faces.push_back({{face[0].vi, face[i].vi, face[i + 1].vi}});
            }
```

File: src/ui/res/di_renderer.ui
Comment (yel-gar) [inline src/ui/res/di_renderer.ui]: Этого тут быть не должно, ошибка с гитом

File: tests/test_core.cpp [hunk 3]
```diff
@@ -102,3 +106,165 @@ TEST(CoreTests, AdvancedMeshRemove) {
...
+    std::vector<di_renderer::math::Vector3> provided_normals = {di_renderer::math::Vector3(0.0F, 0.0F, 1.0F),
+                                                                di_renderer::math::Vector3(0.0F, 0.0F, 1.0F),
+                                                                di_renderer::math::Vector3(0.0F, 0.0F, 1.0F)};
+
+    std::vector<std::vector<di_renderer::core::FaceVerticeData>> faces = {{{0, -1, -1}, {1, -1, -1}, {2, -1, -1}}};
+
+    Mesh mesh(std::move(vertices), {}, std::move(provided_normals), faces);
+
+    ASSERT_EQ(mesh.normals.size(), 3U);
+    EXPECT_NEAR(mesh.normals[0].x, 0.0F, 1e-5F);
```
Comment (yel-gar) [inline tests/test_core.cpp]: Здесь и далее по рекомендации Косенко
```suggestion
    EXPECT_NEAR(mesh.normals[0].x, 0.0F, std::numeric_limits<float>::epsilon());
```

File: src/core/Mesh.hpp [hunk 4]
```diff
@@ -46,11 +47,16 @@ namespace di_renderer::core {
             return normals.size();
         }
         std::size_t face_count() const noexcept {
-            return faces.size();
+            return triangulated_faces.size();
```
Comment (yel-gar) [inline src/core/Mesh.hpp]: Ага, вот почему тест провалился. Просьба вернуть этот метод как был и добавить метод `tri_face_count()` или что-то типа того.

**Review by yel-gar — CHANGES_REQUESTED** [review]
Требуется несколько небольших изменений

**Comment by Pimerroc** [comment]
# Ярослав не вернется..... Вероятно

**Comment by yel-gar** [comment]
@sourcery-ai review
